u2f-ref-code/java/src/com/google/u2f/tools/X509Util.java

```diff
@@ -0,0 +1,151 @@
+package com.google.u2f.tools;
+
+import com.google.u2f.U2FException;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.SignatureException;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * A set of high-level X509 parsing and validation utils
+ */
+public class X509Util {
+  private static final Logger Log = Logger.getLogger(X509Util.class.getName());
+  private static final String X509 = "X.509";
+
+  /**
+   * Parse a single certificate out of a DER-encoded byte array
+   *
+   * @param encodedDerCertificate DER encoded X509 certificate
+   * @return a parsed certificate
+   * @throws CertificateException if parsing could not be done
+   */
+  public static X509Certificate parseCertificate(byte[] encodedDerCertificate)
+      throws CertificateException {
+    return (X509Certificate) CertificateFactory.getInstance(X509).generateCertificate(
+        new ByteArrayInputStream(encodedDerCertificate));
+  }
+
+  /**
+   * Parse a certificate chain out of a DER-encoded byte array.  Will attempt to parse certs
+   * out of the byte array until it can no longer do so.  Will return all the certs (in order) that
+   * were parsed successfully.
+   *
+   * @param encodedDerCertificates DER encoded X509 certificates
+   * @return a parsed certificate chain
+   */
+  public static X509Certificate[] parseCertificateChain(byte[] encodedDerCertificates) {
+    return parseCertificateChain(
+        new DataInputStream(new ByteArrayInputStream(encodedDerCertificates)));
+  }
+
+  /**
+   * Parse a certificate chain out of a DER-encoded DataInputStream.  Will attempt to parse certs
+   * out of the stream until it can no longer do so.  Will return all the certs (in order) that were
+   * parsed successfully.
+   *
+   * @param encodedDerCertificates DER encoded X509 certificates
+   * @return a parsed certificate chain
+   */
+  public static X509Certificate[] parseCertificateChain(DataInputStream encodedDerCertificates) {
+    List<X509Certificate> certChain = new LinkedList<X509Certificate>();
+    try {
+      while (encodedDerCertificates.available() > 0) {
+        certChain.add((X509Certificate) CertificateFactory.getInstance(X509).generateCertificate(
```

Comment:
This wasn't obvious to me at first glance. I think it might be worth adding a comment that CertificateFactory#generateCertificate will create one over a byte stream even when there are trailing bytes remaining. Relying on this allows us to repeatedly call this method over a byte stream containing concatenated certificates.